📖  Add CAPI Provider Operator CAEP (#3833)
kubernetes-sigs/cluster-api: docs/proposals/20201020-management-cluster-operator.md

  * [Additional Details](#additional-details)
     * [Test Plan](#test-plan)
     * [Version Skew Strategy](#version-skew-strategy)
  * [Implementation History](#implementation-history)
  * [Controller Runtime Types](#controller-runtime-types)

## Glossary

The lexicon used in this document is described in more detail
[here](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/reference/glossary.md).
Any discrepancies should be rectified in the main Cluster API glossary.

## Summary

The clusterctl CLI currently handles the lifecycle of Cluster API
providers installed in a management cluster. It provides a great Day 1
experience in getting CAPI up and running. However, clusterctl’s imperative
design makes it difficult for cluster admins to stand up and manage CAPI
management clusters in their own preferred way.

This proposal provides a solution that leverages a declarative API and an
operator to empower admins to handle the lifecycle of providers within the
management cluster

## Motivation

In its current form clusterctl is designed to provide a super simple user
experience for day 1 operations of a Cluster API management cluster with a

> single infrastructure provider.

Author: I've decided to remove the part `with a single infrastructure provider.` since this is false. We can do `clusterctl init -i aws,azure` to install CAPA and CAPZ for example.

upgrade a cluster we now need to supply all the information that was provided
initially during a `clusterctl init` which is inconvenient in many cases such
as distributed teams and CI pipelines where the configuration needs to be
stored and synced externally.

With the management cluster operator, we aim to address these limitations by
introducing an operator that handles the lifecycle of providers within the
management cluster based on a declarative API.

### Goals

- Define an API that enables declarative management of the lifecycle of
  Cluster API providers
- Define and document how providers’ installation and upgrades using the
  operator should work in air-gapped environments.
- Define the lifecycle of the operator itself.
- Identify and document differences between clusterctl CLI and the operator in
  managing the lifecycle of providers, if any.
- Define how the clusterctl CLI should be changed in order to interact with
  the management cluster operator in a transparent and effective way.

### Non-Goals/Future Work

- Deprecate the clusterctl CLI.
- Manage cluster templates using the operator.
- Implement an operator driven version of `clusterctl move`.

> - Manage cert-manager using the operator.

Author: https://github.com/kubernetes-sigs/cluster-api/issues/3996

   ProviderSpec `json:",inline"`
}

// InfrastructureProviderStatus defines the observed state of InfrastructureProvider
type InfrastructureProviderStatus struct {
   ProviderStatus `json:",inline"`
}
```

Below you can find details about `ProviderSpec`,  `ProviderStatus`, which is
shared among all the provider types - Core, Bootstrap, ControlPlane, and
Infrastructure.

```golang
// ProviderSpec defines the desired state of the Provider.
type ProviderSpec struct {
   // Version indicates the provider version.
   // +optional
   Version string `json:"version,omitempty"`

   // Manager defines the properties that can be enabled on the controller manager for the provider.
   // +optional
   Manager ManagerSpec `json:"manager,omitempty"`

   // Deployment defines the properties that can be enabled on the deployment for the provider.
   // +optional
   Deployment DeploymentSpec `json:"deployment,omitempty"`

   // SecretName is the name of the Secret providing the configuration
   // variables for the current provider instance, like e.g. credentials.
   // Such configurations will be used when creating or upgrading provider components.
   // +optional

>    SecretName *string

Reviewer: Let's try to get rid of secret or any credential management. The operator should defer to the providers for any credentials management, if we do find a way to make it global and interface-able we can do that in a future iteration

   // FetchConfig determines how the operator will fetch the components and metadata for the provider.
   // If nil, the operator will try to fetch components according to default
   // settings embedded in the operator and in clusterctl.
   // +optional
   FetchConfig *FetchConfiguration `json:"fetchConfig,omitempty"`

   // Paused prevents the operator from reconciling the provider. This can be
   // used when doing an upgrade or move action manually.
   // +optional
   Paused bool `json:"paused,omitempty"`
}

// ManagerSpec defines the properties that can be enabled on the controller manager for the provider.

> type ManagerSpec struct {

Author: I've done this for ProfilerAddress and FeatureGates. 

It seems that there currently aren't any flags for Verbosity so the provider controller managers will need to include support for a verbosity flag. (_unless I'm missing something_)

Also I realized that for `MaxConcurrentReconciles`, we will need to have logic somewhere (either in the operator or the individual controllers) that can set this value to the various concurrency flags.
That is, for example, for the CAPI provider, this value will need to be set on `cluster-concurrency`, `machine-concurrency`, `machineset-concurrency`, `machinedeployment-concurrency`, `machinepool-concurrency` etc..
If we do this in the operator, that would mean the operator would need to have knowledge of these flags per provider (_which seems like a bad coupling of concerns_). 

We might have to support this flag for each provider and do this within each controller. WDYT?

> type ManagerSpec struct {

Author: Another option would be to remove `MaxConcurrentReconciles` from the API and instead just have users provide the flags as container args if they want to specify a particular `*controller-concurrency` flag.

This would avoid having the need for extra logic in extrapolating how this property maps to a specific provider.
That is, for the `CoreProvider`,  `MaxConcurrentReconciles` needs to map to the above flags but for an `InfrastructureProvider` such as CAPA, it needs to map to `awscluster-concurrency` and `awsmachine-concurrency`.

Personally, I'm leaning to removing this field for the initial implementation and we can add it in if we see the need for an abstraction like this. 
WDYT?

> type ManagerSpec struct {

Author: 👍🏽  If there is a clear user need for something like this, then we should include it. 

Would it be acceptable for the provider controller managers to support a generic `--controller-concurrency` flag that if set, all the other flags such as `awscluster-concurrency` and `awsmachine-concurrency` would be set to that value. This would avoid having the operator to be coupled to the providers. That is, if users set `MaxConcurrentReconciles: 20`, the operator would add `--controller-concurrency=20` to the provider flags.

}
```

Below you can find details about `ProviderSpec`,  `ProviderStatus`, which is
shared among all the provider types - Core, Bootstrap, ControlPlane, and
Infrastructure.

```golang
// ProviderSpec defines the desired state of the Provider.
type ProviderSpec struct {
   // Version indicates the provider version.
   // +optional
   Version *string `json:"version,omitempty"`

   // Manager defines the properties that can be enabled on the controller manager for the provider.
   // +optional
   Manager ManagerSpec `json:"manager,omitempty"`

   // Deployment defines the properties that can be enabled on the deployment for the provider.
   // +optional
   Deployment *DeploymentSpec `json:"deployment,omitempty"`

   // SecretName is the name of the Secret providing the configuration
   // variables for the current provider instance, like e.g. credentials.
   // Such configurations will be used when creating or upgrading provider components.
   // The contents of the secret will be treated as immutable. If changes need
   // to be made, a new object can be created and the name should be updated.
   // The contents should be in the form of key:value.
   // +optional

>    SecretName *string

Author: Just adding this note. We are currently discussing the need for this `SecretName` field. 

The reason we have a secret is so that we can store credentials that may be required by the providers as part of the InfrastructureProvider setup. For example, `AWS_B64ENCODED_CREDENTIALS` or `AZURE_CLIENT_SECRET_B64`. However, if we remove the need for such sensitive information as part of declaring an `InfrastructureProvider`, then all we have left are variables like `AWS_REGION`  or `AZURE_ENVIRONMENT`, which are not considered sensitive.

So we can instead replace the `SecretName` to something like `Variables map[string][string]` and these variables can be used to evaluate the `infrastructure-components.yaml` for the specified Infra provider.

There are on-going discussions with the infra providers regarding this change and the need to have a separate entity to manage/retrieve the identity required by the management providers to manage workload clusters.

See this proposal: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/proposal/20200506-single-controller-multitenancy.md

Author: https://github.com/kubernetes-sigs/cluster-api/issues/3881#issuecomment-743370519

Author: Happy New Year Ping!
Just wanted to raise this concern and involve provider authors to provide perspective on managing credentials for the CAPI components.

Author: @yastij The scope of the credentials usage will be the same as it is today (pre-operator). That is, it seems for CAPA it is used to just bootstrap an IAM instance profile.

Ok...so it seems for now we'll leave the `SecretName` as is. 
This will decouple the single-controller-multitenancy work and associated credential management from this CAEP. Also it won't feel like we are rushing the credential management design in order to fulfill this CAEP.

WDYT?

Reviewer 2: +1

   ctrlruntime.ControllerManagerConfigurationSpec `json:",inline"`

   // ProfilerAddress defines the bind address to expose the pprof profiler (e.g. localhost:6060).
   // Default empty, meaning the profiler is disabled.
   // Controller Manager flag is --profiler-address.
   // +optional
   ProfilerAddress *string `json:"profilerAddress,omitempty"`

   // MaxConcurrentReconciles is the maximum number of concurrent Reconciles
   // which can be run. Defaults to 10.
   // +optional
   MaxConcurrentReconciles *int `json:"maxConcurrentReconciles,omitempty"`

   // Verbosity set the logs verbosity. Defaults to 1.
   // Controller Manager flag is --verbosity.
   // +optional
   Verbosity int `json:"verbosity,omitempty"`

   // Debug, if set, will override a set of fields with opinionated values for
   // a debugging session. (Verbosity=5, ProfilerAddress=localhost:6060)
   // +optional
   Debug bool `json:"debug,omitempty"`

   // FeatureGates define provider specific feature flags that will be passed
   // in as container args to the provider's controller manager.
   // Controller Manager flag is --feature-gates.

>    FeatureGates map[string]bool `json:"featureGates, omitempty"`

Author: @CecileRobertMichon @detiber When y'all get the chance, please take a look at @vincepri's comment above.

   // The same holds for `ManagerSpec.FeatureGates` and `--feature-gates`.
   // +optional
   Args map[string]string `json:"args,omitempty"`

   // Compute resources required by this container.
   // +optional
   Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
}

// ImageMeta allows to customize the image used
type ImageMeta struct {
    // Repository sets the container registry to pull images from.
    // +optional
    Repository *string `json:"repository,omitempty`

    // Name allows to specify a name for the image.
    // +optional
    Name *string `json:"name,omitempty`

    // Tag allows to specify a tag for the image.
    // +optional
    Tag *string `json:"tag,omitempty`
}

// FetchConfiguration determines the way to fetch the components and metadata for the provider.
type FetchConfiguration struct {
   // URL to be used for fetching the provider’s components and metadata from a remote Github repository.
   // For example, https://github.com/{owner}/{repository}/releases
   // The version of the release will be `ProviderSpec.Version` if defined
   // otherwise the `latest` version will be computed and used.

>    // +optional

Reviewer 2: Is it required for the components to be fetched from a github release or could they be stored somewhere else?

>    // +optional

Author: As per the FetchConfiguration, they can either be a Github release (URL) or a ConfigMap (Selector).

As part of the initial work, we are trying to be in parity with the existing clusterctl behavior. That is, the URL must have a scheme of `https` and a host of `github.com`.

https://github.com/kubernetes-sigs/cluster-api/blob/b6c1b5b3db3d062cf89e7c66ab233fe503767362/cmd/clusterctl/client/repository/client.go#L161-L168

We can add support for fetching components for other sources in the future. Or we can create a better abstraction here for supporting other FetchConfiguration sources. WDYT?

---
title: Management Cluster Operator
authors:
  - "@fabriziopandini"
  - "@wfernandes"
reviewers:
  - "@vincepri"
  - "@ncdc"
  - "@justinsb"
  - "@detiber"
  - "@CecileRobertMichon"
creation-date: 2020-09-14
last-updated: 2020-10-20
status: implementable (aspirational :wink:)
see-also:
https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191016-clusterctl-redesign.md
---

> # Management cluster operator

Reviewer 2: I've lost access to the google doc with my previous comment but what did we conclude on the phrasing "management cluster" operator vs. "providers"?

Author: I seem to have lost access to that doc as well. I think @fabriziopandini might be the owner of the document.
But I remember providing some suggestions although that was a while ago 😄 . 
I'm not too attached to the name of this operator and I'm open to suggestions.

I believe the argument was since this essentially operating on the providers, it should be called like  _Provider Operator_. As per our Glossary, the definition of Management Cluster is _The cluster where one or more Infrastructure Providers run, and where resources (e.g. Machines) are stored. Typically referred to when you are provisioning multiple workload clusters._ 
From this definition, this operator is not necessarily operating on the objects stored within the management clusters but rather just the providers.

So I'm ok with calling it the **CAPI Provider Operator**. I'm open to suggestions from others 🙂  WDYT?

Reviewer 2: I like "CAPI Provider Operator"

what do others think?

Reviewer 3: I'm ok with CAPI Provider Operator

Reviewer: Technically we also manage Cluster API itself, but we can see it as a provider? 😄

Reviewer 2: It's consistent with the naming of the types, CAPI is a "CoreProvider"

for day 1 operations of a Cluster API management cluster.

However such design is not optimized for supporting declarative approaches
when operating Cluster API management clusters.

These declarative approaches are important to enable GitOps workflows in case
users don't want to rely solely on the `clusterctl` CLI.

Providing a declarative API also enables us to leverage controller-runtime's
new component config and allow us to configure the controller manager and even
the resource limits of the provider's deployment.

Another example is improving cluster upgrades. In order to upgrade a cluster
we now need to supply all the information that was provided initially during a
`clusterctl init` which is inconvenient in many cases such as distributed
teams and CI pipelines where the configuration needs to be stored and synced
externally.

With the management cluster operator, we aim to address these use cases by
introducing an operator that handles the lifecycle of providers within the
management cluster based on a declarative API.

### Goals

- Define an API that enables declarative management of the lifecycle of
  Cluster API and all of its providers.

> - Support air-gapped environments through sufficient documentation initially.

Author: Currently, as part of good UX, clusterctl reaches out to Github repositories to download the infrastructure-components.yaml and the cluster-template.yaml for the providers and the workload clusters. 

We are simply stating that we will be providing sufficient documentation to show how the current design will also enable current support for Air Gapped environments. (also example usage in Air Gapped environment section)

- Identify and document differences between clusterctl CLI and the operator in
  managing the lifecycle of providers, if any.
- Define how the clusterctl CLI should be changed in order to interact with
  the management cluster operator in a transparent and effective way.
- To support the ability to upgrade from a v1alpha3 based version (v0.3.[TBD])
  of Cluster API to one managed by the operator.

### Non-Goals/Future Work

- Implement an operator driven version of `clusterctl move`.
- Manage cert-manager using the operator.

> - Support multiple installations of the same provider within a management

Reviewer 4: Coming from the discussion yesterday again:
Not supporting it here should be fine but would it still be possible?
As far as I can tell, there is nothing directly prohibiting me from having 2 installations of the same provider in different namespaces here.

Reviewer: That's correct, the operator itself doesn't stop any user to be able to run multiple controllers in a management cluster, but it doesn't/won't support it natively for the v1alpha4 iteration.

The Cluster API community has been going in this direction because it's a hard problem to support version skews in the same management cluster with a generic operator. The goal of this operator proposal is to fulfill the 80% use case: run a management cluster with a simple deployment. Meanwhile, we should make the 20% possible (same as every other feature we usually build).

Author: We may need to revise these statements in the _Installing a Provider_ section

https://github.com/kubernetes-sigs/cluster-api/pull/3833/files#diff-a8b3fa322e4449932bb3e414c2582ca1acae4ae38e9fd9036e3df433f5b5528bR612-R615

> Before installing any provider following preflight checks will be executed :
>- There should not be another instance of the same provider (same Kind, same name) in any namespace.
>- The Cluster API contract the provider is abiding by, e.g. v1alpha4, must match the contract of the core provider.

Reviewer 3: I think we should keep these checks in the operator because it should ensure the consistency of the cluster no matter if it is used with clusterctl or not.

Reviewer: The checks should still be in place, but we shouldn't disallow running more than one controller in Cluster API code. The difference here is that the use cases supported by the operator are on a smaller scale, while we allow others to run different deployments if they need to.

  cluster in light of [issue 3042] and [issue 3354].
- Support any template processing engines.
- Support the installation of v1alpha3 providers using the operator.

## Proposal

### User Stories

1. As an admin, I want to use a declarative style API to operate the Cluster
   API providers in a management cluster.
1. As an admin, I would like to have an easy and declarative way to change
   controller settings (e.g. enabling pprof for debugging).
1. As an admin, I would like to have an easy and declarative way to change the
   resource requirements (e.g. such as limits and requests for a provider
   deployment).
1. As an admin, I would like to have the option to use clusterctl CLI as of
   today, without being concerned about the operator.

> 1. As an admin, I would like to be able to install the operator using kubectl

Author: Yes. I'll try and find a place in the caep to make this clear.

> 1. As an admin, I would like to be able to install the operator using kubectl

Author: I added a section for Clusterctl in the Implementation Details section below.

   apply, without being forced to use clusterctl.

### Implementation Details/Notes/Constraints

#### Existing API Types Changes

The existing `Provider` type used by the clusterctl CLI will be deprecated and
its instances will be migrated to instances of the new API types as defined in
the next section.

The management cluster operator will be responsible for migrating the existing
provider types to support GitOps workflows excluding `clusterctl`.

#### New API Types

These are the new API types being defined.

There are separate types for each provider type - Core, Bootstrap,
ControlPlane, and Infrastructure. However, since each type is similar, their
Spec and Status uses the shared types - `ProviderSpec`,  `ProviderStatus`
respectively.

If you prefer to see how the API can be used instead of reading the type
definition feel free to jump to the [Example API Usage
section](#example-api-usage)

```golang
// CoreProvider is the Schema for the CoreProviders API

> type CoreProvider struct {

Author: So we actually did define separate Specs for each provider type but since they were using all using the `ProviderSpec`, the suggestion was to remove the specific `CoreProviderSpec` and just have it as `ProviderSpec`. In the future, if we need to include specific properties we can embed the `ProviderSpec` type and this will not break compatibility.
```golang
type CoreProviderSpec struct {
  ProviderSpec
  AnotherField string
}
```

For the reasoning for having separate types like `CoreProvider`, `BootstrapProvider` etc. this is because we still need a way to identify the providers. Doing something like below doesn't seem to follow k8s patterns _at least IMO_.
```golang
type Provider struct {
  Type string
}
```

> type CoreProvider struct {

Reviewer 2: sgtm

   // +optional
   Affinity *corev1.Affinity `json:"affinity,omitempty"`

   // List of containers specified in the Deployment
   // +optional
   Containers []ContainerSpec `json:"containers"`
}

// ContainerSpec defines the properties available to override for each
// container in a provider deployment such as Image and Args to the container’s
// entrypoint.
type ContainerSpec struct {
   // Name of the container. Cannot be updated.
   Name string `json:"name"`

   // Container Image Name
   // +optional
   Image *ImageMeta `json:"image,omitempty"`

   // Args represents extra provider specific flags that are not encoded as fields in this API.
   // Explicit controller manager properties defined in the `Provider.ManagerSpec`
   // will have higher precedence than those defined in `ContainerSpec.Args`.
   // For example, `ManagerSpec.SyncPeriod` will be used instead of the
   // container arg `--sync-period` if both are defined.
   // The same holds for `ManagerSpec.FeatureGates` and `--feature-gates`.
   // +optional
   Args map[string]string `json:"args,omitempty"`

   // List of environment variables to set in the container.
   // +optional

>    Env []corev1.EnvVar `json:"env,omitempty"

Author: I added `Env` to explicitly support environment variables to the container. However as I was looking at the corev1 api I noticed the following comments and tags and I was curious if we need to specify these tags as well.

```
// +patchMergeKey=name
// +patchStrategy=merge
Env []EnvVar `json:"env,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,7,rep,name=env"`
```

Reviewer: We should try them out, IIRC latest controller runtime/tools versions should support these, although it needs to be tested

Author: I'll leave out the annotations from the proposal for now and we can try adding them in when we start implementing.

  expected to contain components and metadata for a specific version only. So
  if multiple versions of the providers need to be specified, they can be
  added as separate ConfigMaps and labeled with the same selector. This
  provides the same behavior as the “local” provider repositories but now from
  within the management cluster.
- `FetchConfiguration` is used only during init and upgrade operations.
  Changes made to the contents of `FetchConfiguration` will not trigger a
  reconciliation. This is similar behavior to `ProviderSpec.SecretName`.

**Validation Rules for ProviderSpec.ManagerSpec**
- The ControllerManagerConfigurationSpec is a type from
  `controller-runtime/pkg/config` and is an embedded into the `ManagerSpec`.
  This type will expose LeaderElection, SyncPeriod, Webhook, Health and
  Metrics configurations.
- If `ManagerSpec.Debug` is set to true, the operator will not allow changes
  to other properties since it is in Debug mode.
- If you need to set specific concurrency values for each reconcile loop (e.g.
  `awscluster-concurrency`), you can leave
  `ManagerSpec.MaxConcurrentReconciles` nil and use `Container.Args`.

> - If `ManagerSpec.MaxConcurrentReconciles` is set and a specific concurrency

Reviewer 2: I like this logic, `MaxConcurrentReconciles` is the overall concurrency but any specific flag tuning takes precedence

Author: Yup. As mentioned above, I suggest this be deferred to the individual controller level. That is, if the user has set `MaxConcurrentReconciles` and also defined a specific flag, then the operator can pass to the controller `--controller-concurrency=20 --awscluster-concurrency=10` and the controller manager can override `awscluster-concurrency` accordingly.

Apologies if this is getting too "implementationy".

  container arg `sync-period`. This is true also for
  `ManagerSpec.FeatureGates`, that is, it will have higher precedence to the
  container arg `feature-gates`.
- If no `ContainerSpec.Resources` are defined, the defaults on the Deployment
  object within the provider’s components yaml will be used.

**Open Questions**
1. Should the CRDs be global objects or namespaced objects?
    A provider lives in a namespace (e.g. capa-system), so it is nice to have
    the object namespaced and use metadata.namespace. However, with the new
    approach to multitenancy, we should enforce are like singleton (a
    provider, identified by Kind & name, should be unique at cluster level)
    It will be nice to provide a nice UX around kubectl get <providers>
1. What should the Group/Version for this API be?
    Currently, we have it as `management.cluster.x-k8s.io/v1alpha1`. Should
    the version be `v1alpha4` to match the CAPI version it is being introduced
    in? This would be consistent with what we do for experiments.
1. Should we have multiple types (i.e. CoreProvider, BootstrapProvider,
   ControlPlaneProvider and InfrastructureProvider) vs. a single Provider
   type?

>    Even though we can use CRD categories to group them, the output may

Reviewer 2: ah this is the same question I asked above :)

 deployment:
   containers:
   - name: manager
     args:
         # These are controller flags that are specific to a provider; usage
         # is reserved for advanced scenarios only.
         awscluster-concurrency: 12
         awsmachine-concurrency: 11
```

2. As an admin, I want to install aws infrastructure provider but override
   the container image of the CAPA deployment.

```yaml
---
apiVersion: management.cluster.x-k8s.io/v1alpha1
kind: InfrastructureProvider
metadata:
 name: aws
 namespace: capa-system
spec:
 version: v0.6.0
 secretName: aws-variables
 deployment:
   containers:
   - name: manager
     image: gcr.io/myregistry/capa-controller:v0.6.0-foo
```

3. As an admin, I want to change the resource limits for the manager pod in
   my control plane provider deployment.

```yaml
---
apiVersion: management.cluster.x-k8s.io/v1alpha1
kind: ControlPlaneProvider
metadata:
 name: kubeadm
 namespace: capi-kubeadm-control-plane-system
spec:
 version: v0.3.10
 secretName: capi-variables
 deployment:
   containers:
   - name: manager
     resources:
       limits:
         cpu: 100m
         memory: 30Mi
       requests:
         cpu: 100m
         memory: 20Mi
```

> 4. As an admin, I would like to fetch my azure provider components from a

Reviewer 2: would be good to include an example of an infrastructure provider with no credentials as well

Author: I can add one in there but that would just be omitting the `secretName` field since it's optional anyways. But I'll look for an example, I don't think it hurts to have more examples. 🙂

following steps:
- Fetching the provider artifacts (the components yaml and the metadata.yaml
  file)
- Applying image overrides, if any
- Replacing variables
- Applying the resulting yaml to the cluster

As a final consideration, please note that
- The operator executes installation for 1 provider at time, while `clusterctl
  init` manages installation of a group of providers with a single operation.
- `clusterctl init` uses environment variables and a local configuration file,
  while the operator uses a Secret; given that we want the users to preserve
  current behaviour in clusterctl, the init operation should be modified to
  transfer local configuration to the cluster.
  As part of `clusterctl init`, it will obtain the list of variables required
  by the provider components and read the corresponding values from the config
  or environment variables and build the secret.
  Any image overrides defined in the clusterctl config will also be applied to
  the provider's components.

##### Upgrading a provider

In order to trigger an upgrade of a new Cluster API provider you have to
change the `spec.Version` field.

> Upgrading a provider in the management cluster must abide by the golden rule

Reviewer 2: is there a specific order in which providers should be upgraded?

Author: We will be following the same ruleset that clusterctl follows. 

I'm assuming that the CoreProvider must go first. We have that rule specified above as part of the install process. That is, when setting up the management cluster, the CoreProvider must be installed first and the other providers will be re-queued until it has been installed. (see Installing a Provider section above)

object.

Deletion of the provider will be blocked if any workload cluster using the
provider still exists.

Additionally, deletion of a core provider should be blocked if there are still
other providers in the management cluster.

#### Upgrade from v1alpha3 management cluster to v1alpha4 cluster

Cluster API will provide instructions on how to upgrade from a v1alpha3
management cluster, created by clusterctl to the new v1alpha4 management
cluster. These operations could require manual actions.

Some of the actions are described below:
- Run webhooks as part of the main manager. See [issue 3822].

More details will be added as we better understand what a v1alpha4 cluster
will look like.

#### Operator Lifecycle Management

##### Operator Installation

- `clusterctl init` will install the operator and its corresponding CRDs as a
  pre-requisite if the operator doesn’t already exist. Please note that this
  command will consider image overrides defined in the local clusterctl config
  file.
- If the admin does not want to use clusterctl to install the operator, it is
  possible to kubectl apply the operator yaml that will be published in the
  cluster-api release artifacts.

> - There can be only one operator installed per management cluster, across all

Reviewer 2: how is this enforced?

Author: Good point. `clusterctl` will enforce this if the users use `clusterctl init`. 
If the user installs the operator manually (kubectl), then we rely on documentation? 🤷🏽 
🤔  I guess the operator itself could do some sort of check. That is, if installing the second instance of the operator, it can determine if there is an operator already installed. Not sure about the implementation. But I'm open to suggestions.

Reviewer 3: I'm +1 to remove this sentence given that it could be confusing.

We are going to support many running replicas of the operator, but only one will be active via leader election (e.g for HA purposes).

What is open for discussion is if this operator should support the --namespace flag (for reconciling objects in a single namespace), but this is strictly linked to the fact of using global or namespaced CRDs.

Author: Ok. I'll remove this line since it is a little confusing.

  * [Implementation History](#implementation-history)

## Glossary

The lexicon used in this document is described in more detail
[here](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/reference/glossary.md).
Any discrepancies should be rectified in the main Cluster API glossary.

## Summary

The clusterctl CLI currently handles the lifecycle of Cluster API
providers installed in a management cluster. It provides a great Day 1
experience in getting CAPI up and running. However, clusterctl’s imperative
design could make it difficult for cluster admins to stand up and manage CAPI
management clusters in their own preferred way.

This proposal provides a solution that leverages a declarative API and an
operator to empower admins to handle the lifecycle of providers within the
management cluster

## Motivation

In its current form clusterctl is designed to provide a super simple user
experience for day 1 operations of a Cluster API management cluster with a
single infrastructure provider.

However such design implicitly is not optimized for supporting declarative
approaches when operating Cluster API management clusters (e.g. replicating

> clusterctl init is complex).  With the management cluster operator, we aim to

Reviewer 3: What about:

However such design does not work well with declarative approaches, because the integration of the clusterctl CLI in a pipeline triggered by git commits could be problematic, and without the CLI the users are losing clusterctl advantages like e.g. the logic for fetching latest version of a provider embedded in clusterctl init.

> What is the word "implicitly" doing here?
> What does it mean to optimize for declarative approaches?

What about ""

  the management cluster operator in a transparent and effective way.

### Non-Goals/Future Work

- Deprecate the clusterctl CLI
- Manage cluster templates using the operator
- Implement an operator driven version of `clusterctl move`
- Manage cert-manager using the operator
- Support multiple installations of the same provider within a management
  cluster in light of [issue
  3042](https://github.com/kubernetes-sigs/cluster-api/issues/3042) and [issue
  3354](https://github.com/kubernetes-sigs/cluster-api/issues/3354)

## Proposal

### User Stories

1. As an admin, I want to use a declarative style API to operate the Cluster
   API providers in a management cluster.
1. As an admin, I would like to have an easy and declarative way to change
   controller settings (e.g. enabling pprof for debugging).
1. As an admin, I would like to have an easy and declarative way to change the
   resource requirements such as limits and requests for a provider
   deployment.
1. As an admin, I would like to upgrade the Cluster API providers in a
   management cluster without having to provide the same information I used
   when I did `clusterctl init`.

> 1. As an admin, I would like to have the option to use clusterctl CLI as of

Reviewer 5: Is this saying that users can continue to use clusterctl without the operator, or just that they can be unaware it?

Author: Moving forward the operator will be required by clusterctl in order to operate. However, the user need not be aware of the existence of the operator. As defined in the Operator Installation, when a user now does a `clusterctl init` on a clean cluster, it will install the operator and then perform similar operations as today.

   // +optional
   Manager ManagerSpec `json:"manager,omitempty"`

   // Deployment defines the properties that can be enabled on the deployment for the provider.
   // +optional
   Deployment DeploymentSpec `json:"deployment,omitempty"`

   // SecretName is the name of the Secret providing the configuration
   // variables for the current provider instance, like e.g. credentials.
   // Such configurations will be used when creating or upgrading provider components.
   // +optional
   SecretName *string

   // FetchConfig determines how the operator will fetch the components and metadata for the provider.
   // If nil, the operator will try to fetch components according to default
   // settings embedded in the operator and in clusterctl.
   // +optional
   FetchConfig *FetchConfiguration `json:"fetchConfig,omitempty"`

   // Paused prevents the operator from reconciling the provider. This can be
   // used when doing an upgrade or move action manually.
   // +optional
   Paused bool `json:"paused,omitempty"`
}

// ManagerSpec defines the properties that can be enabled on the controller manager for the provider.

> type ManagerSpec struct {

Reviewer 3: I think it is worth to put some focus on the UX around feature flags because they are currently managed as a sort of exception given that currently, they are the only element of a manager spec that can be influenced by the users during init (using variables).
Also, in the current clusterctl, the same feature flag variable applies to many providers

Author: I included a property for `FeatureGates` in the `ManagerSpec` and clarified its behavior in the validation rules for `ContainerSpec.Args`.

   // ControllerManagerConfigurationSpec defines the desired state of GenericControllerManagerConfiguration.

>    v1alpha1.ControllerManagerConfigurationSpec

Reviewer 3: might be use a more expressive alias (I guess this comes from controller runtime).
Also, for the sake of documentation, might be worth to import a copy of this type in the doc

   MaxConcurrentReconciles *int `json:"maxConcurrentReconciles,omitempty"`

   // Verbosity set the logs verbosity. Defaults to 1.
   // +optional
   Verbosity int `json:"verbosity,omitempty"`

   // Debug, if set, will override a set of fields with opinionated values for
   // a debugging session. (Verbosity=5, ProfilerAddress=localhost:6060)
   // +optional
   Debug bool `json:"debug, omitempty"`
}

// DeploymentSpec defines the properties that can be enabled on the Deployment for the provider.
type DeploymentSpec struct {
   // Number of desired pods. This is a pointer to distinguish between explicit zero and not specified. Defaults to 1.
   // +optional
   Replicas int `json:"replicas,omitempty"`

   // List of containers specified in the Deployment
   // +optional
   Containers []ContainerSpec `json:"containers"`
}

// ContainerSpec defines the properties available to override for each
// container in a provider deployment such as Image and Args to the container’s
// entrypoint.
type ContainerSpec struct {
   // Name of the container. Cannot be updated.
   Name string `json:"name"`

   // Docker image name
   // +optional
   Image string `json:"image,omitempty"`

>    // Args represents extra provider specific flags that are not encoded as fields in this API.

Author: We mentioned it in the validation rules for the `ContainerSpec`.

   // +optional
   Args map[string]string `json:"args,omitempty"`

   // Compute resources required by this container.
   // +optional
   Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
}

// FetchConfiguration determines the way to fetch the components and metadata for the provider.

> type FetchConfiguration struct {

Author: We can add it in the Validation rules section for FetchConfiguration.

   // +optional
   URL *string `json:"url,omitempty"`

   // Selector to be used for fetching provider’s components and metadata from
   // ConfigMaps stored inside the cluster. Each ConfigMap is expected to contain
   // components and metadata for a specific version only.
   // +optional
   Selector *metav1.LabelSelector `json:"selector,omitempty"`
}

// ProviderStatus defines the observed state of the Provider.
type ProviderStatus struct {
   // Contract will contain the core provider contract that the provider is
   // abiding by, like e.g. v1alpha3.
   Contract string `json:"contract,omitempty"`

   // Conditions define the current service state of the cluster.
   // +optional
   Conditions Conditions `json:"conditions,omitempty"`

   // ObservedGeneration is the latest generation observed by the controller.
   // +optional
   ObservedGeneration int64 `json:"observedGeneration,omitempty"`
}
```

**Validation and defaulting rules for Provider and ProviderSpec**
- The `Name` field within `metav1.ObjectMeta` could be any valid Kubernetes
  name; however, it is recommended to use Cluster API provider names. For

>   example, aws, vsphere, kubeadm.

Author: Good point. I'll mention that here again. We mentioned the behavior in the Validation rules for the FetchConfiguration but it doesn't hurt to add that here as well.

- `ProviderSpec.Version` should be a valid default version with the "v" prefix
  as commonly used in the Kubernetes ecosystem; if this value is nil when a
  new provider is created, the operator will determine the version to use
  applying the same rules implemented in clusterctl (latest).
- The content of `ProviderSpec.SecretName` will be treated as immutable. If
  changes need to be made, a new object can be created and the name should be
  updated.
- The content of `ProviderSpec.SecretName` should be in the form of key:value.
- The `ProviderSpec.SecretName` specified must be in the same namespace as the
  provider.
- When a provider will be paused the number of replicas will be scaled to 0;
  the operator adds a new
  `management.cluster.x-k8s.io/original-controller-replicas` annotations to

>   store the original replica count.

Author: I'll add an example for pausing declaratively and then move this note to that section.

  `controller-runtime/pkg/config` and is an embedded into the `ManagerSpec`.
  This type will expose LeaderElection, SyncPeriod, Webhook, Health and
  Metrics configurations.
- If `ManagerSpec.Debug` is set to true, the operator will not allow changes
  to other properties since it is in Debug mode.
- If you need to set specific concurrency values for each reconcile loop (e.g.
  `awscluster-concurrency`), you can leave
  `ManagerSpec.MaxConcurrentReconciles` nil and use `Container.Args`.
- If `ManagerSpec.MaxConcurrentReconciles` is set and a specific concurrency
  flag such as `awscluster-concurrency` is set on the `Container.Args`, then
  the more specific concurrency flag will have higher precedence.

**Validation Rules for ContainerSpec**
- The `ContainerSpec.Args` will ignore the key `namespace` since the operator
  enforces a deployment model where all the providers should be configured to
  watch all the namespaces.
- Explicit controller manager properties defined in the `Provider.ManagerSpec`
  will have higher precedence than those defined in `ContainerSpec.Args`. That
  is, if `ManagerSpec.SyncPeriod`  is defined it will be used instead of the

>   container arg `sync-period`.

Author: That is one option but it feels restrictive because now the user will need to look up the API and see what they can define in the ManagerSpec vs what they can define in the ContainerSpec.Args.
I thought that it would be better to just have this rule so that we can just do a union of all the flags that need to be set on the controller manager and use the one defined in ManagerSpec if there are duplicates.

>   container arg `sync-period`.

Reviewer 3: I have no strong opinions, but I personally prefer to get an explicit validation error message when creating the object instead of trying to find out ex-post what is the merge rule

   specific controller flags.

```yaml
apiVersion: v1
kind: Secret
metadata:
 name: aws-variables
 namespace: capa-system
type: Opaque
data:
 AWS_REGION: ...
 AWS_ACCESS_KEY_ID: ...
 AWS_SECRET_ACCESS_KEY: ...
---
apiVersion: management.cluster.x-k8s.io/v1alpha1
kind: InfrastructureProvider
metadata:
 name: aws
 namespace: capa-system
spec:
 version: v0.6.0
 secretName: aws-variables
 manager:
   # These top level controller manager flags, supported by all the providers.
   # These flags come with sensible defaults, thus requiring no or minimal
   # changes for the most common scenarios.
   metricsAddress: ":8181"
   syncPeriod: 660
 fetchConfig:
   url: https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases
 deployment:
   containers:
   - name: manager
     args:
         # These are controller flags that are specific to a provider; usage
         # is reserved for advanced scenarios only.
         awscluster-concurrency: 12
         awsmachine-concurrency: 11
```

2. As an operator, I want to install aws infrastructure provider but override
   the container image of the CAPA deployment.

```yaml

> <-- Secret removed to reduce verbosity -->

Reviewer 5: I think you can just omit this completely if you want to reduce verbosity :)
```suggestion
```

metadata:
 name: azure
 namespace: capz-system
spec:
 version: v0.4.9
 secretName: azure-variables
 fetchConfig:
   url: https://github.com/myorg/awesome-azure-provider/releases

```

5. As an operator, I would like to fetch my azure provider components from
   within the cluster because I’m working within an air-gapped environment.

See example in the [air-gapped environment section](#air-gapped-environment)

#### Operator Behaviors

##### Installing a provider

In order to install a new Cluster API provider with the management cluster
operator you have to kubectl apply the following yaml:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: aws-configuration
  namespace: capa-system
type: Opaque
data:
  AWS_REGION: ...
  AWS_ACCESS_KEY_ID: ...
  ...
---
apiVersion: management.cluster.x-k8s.io/v1alpha4
kind: InfrastructureProvider
metadata:
  name: aws
  namespace: capa-system
spec:
 configRef:
   apiVersion: v1
   kind: Secret
   name: aws-configuration
```

When processing a Provider object the operator will apply the following rules.

- Providers with  `spec.Type == CoreProvider` will be installed first; the
  other providers will be requeued until the core provider exists.

> - Before installing any provider following preflight checks will be executed :
> - There should not be another instance of the same provider (same Kind, same
>   name) in any namespace.
> - The Cluster API contract the provider is abiding by, e.g. v1alpha3, must
>   match the contract of the core provider.
> - The operator will set conditions on the Provider object to surface any
>   installation issues such as pre-flight checks and/or order of installation
>   to accurately inform the user.

Reviewer 5: ```suggestion
- Before installing any provider following preflight checks will be executed:
  - There should not be another instance of the same provider (same Kind, same
  name) in any namespace.
  - The Cluster API contract the provider is abiding by, e.g. v1alpha3, must
  match the contract of the core provider.
  - The operator will set conditions on the Provider object to surface any
  installation issues such as pre-flight checks and/or order of installation
  to accurately inform the user.
```

`spec.Paused` field to true for each provider; the operator will block any
contract upgrade until all the providers are paused.

After all the providers are in paused state, you can proceed with the upgrade
as described in the previous paragraph (change the `spec.Version` field).

Once all the providers are upgraded to a version that abides to the new
contract, it is possible for the operator to unpause providers; the operator
does not allow to unpause providers if there are still providers abiding to
the old contract.

Please note that we are planning to embed this sequence (pause - upgrade -
unpause) as a part of `clusterctl upgrade apply` command when there is a
contract change.

##### Changing a provider

On top of changing a provider version (upgrades), the operator supports also
changing other provider fields, most notably controller flags and variables.

The operation internally works like upgrades: The current instance of the
provider is deleted, while preserving CRDs, namespaced and user objects A new
instance of the provider is installed with the new set of flags/variables.

Please note that clusterctl currently does not support this operation.

> See Example 1 in [Example API Usage](#example-api-usage)

Reviewer 3: what about expanding this a little by explaining that the user is required to kubectl edit (or to kubeactl apply again) the object describing the provider

##### Deleting a provider

In order to delete a provider you have to delete the corresponding provider
object.

Deletion of the provider will be blocked if any workload cluster using the
provider still exists.

Additionally, deletion of a core provider should be blocked if there are still
other providers in the management cluster.

#### Operator Lifecycle Management

##### Operator Installation

- `clusterctl init` will install the operator and its corresponding CRDs as a
  pre-requisite if the operator doesn’t already exist. Please note that this
  command will consider image overrides defined in the local clusterctl config
  file.
- If the admin does not want to use clusterctl to install the operator, it is
  possible to kubectl apply the operator yaml that will be published in the
  cluster-api release artifacts.
- There can be only one operator installed per management cluster, across all
  namespaces.

##### Operator Upgrade

- The admin can use `clusterctl upgrade operator` to upgrade the operator
  components. Please note that this command will consider image overrides

>   defined in the local clusterctl config file.

Reviewer 3: I think we should make this also part of the `clusterctl upgrade plan` and part of the `clusterctl upgrade apply` workflow similarly to what we are doing for cert-manager.
This does not exclude the idea for providing separated commands for upgrading the operator alone (or cert-manager), but I would like clusterctl to always provide an easy button for all the operations.

Author: Make sense. Since we currently check to upgrade cert-manager as part of `upgrade plan` and `upgrade apply` we can do the same for the operator.

- If the admin doesn’t want to use clusterctl, they can use kubectl apply with
  the latest version of the operator yaml that will be published in the
  cluster-api release artifacts.
- clusterctl will require a matching operator version. In the future, when
  clusterctl move to beta/GA, we will reconsider supporting version skew
  between clusterctl and the operator.

##### Operator Delete

- clusterctl will delete the operator as part of the `clusterctl delete --all`
  command.
- If the admin doesn’t want to use clusterctl, they can use kubectl delete.
  However, it’s the admin’s responsibility to verify that there are no

>   providers running in the management cluster.

Reviewer 3: What if the admin tries to use clusterctl CLI after the operator is deleted? should we document this here?

Author: I'm assuming that if the user is using the latest clusterctl (_one which uses the operator_) after the operator is deleted, the clusterctl will respond with an error saying that it is unable to detect the operator and that the operator needs to be installed.
Unless of course, the user does a `clusterctl init`, after which it will install the operator again.

Reviewer 3: makes sense

#### Air gapped environment

> In order to use the operator in an air-gapped environment it is required

Reviewer 3: ```suggestion
In order to install Cluster API providers in an air-gapped environment using the operator, it is required to address the following problems
```

1. Make the operator work in air-gapped environment

>    - To provide image overrides for the operator itself.

Reviewer 3: ```suggestion
   - To provide image overrides for the operator itself in order to pull the images
     from an accessible image repository.
```

Reviewer 3: Are you envisioning using the current clusterctl image overrides for this?
Also, does this will setting will be local to the admin workstation?

Author: So in the first point of the [Operator Installation section](https://github.com/kubernetes-sigs/cluster-api/pull/3833/files#diff-a8b3fa322e4449932bb3e414c2582ca1acae4ae38e9fd9036e3df433f5b5528bR699-R702), we did mention that we are going to use the imageOverrides section for this.
 
But I do see your point that would cause another "local configuration" file that would need to be maintained separately. This feels like a chicken/egg problem here.
🤔 

So there are a couple of options that come to mind.
- The admin uses their own modified `operator.yaml` to install and upgrade the operator. This `operator.yaml` can be used within their own gitops practices this way. This can be viable but then things like `clusterctl upgrade plan` won't be able to inform if an upgrade is available or not.
- Currently we have a providers configuration in our `clusterctl.yaml` along with `imageOverrides` section, which we don't need anymore because of the declarative nature of our Provider API. However, since we still have other specific configs like the overrides layer, cert-manager-timeout and debug/log level there is still a need for `clusterctl.yaml` config. Maybe we can include an operator section. This section would be similar to the providers configuration in the sense that the admin can provide a URL to clusterctl from where to obtain the operator artifacts. This URL can point to a repository within the air-gapped environment and is under the control of the admins so that they can make the image overrides to that version directly. However, this feels a bit more round-about of the first option.

I'm leaning towards keeping it simple by using the imageOverrides defined in the local clusterctl config for now and documenting that users can maintain the `operator.yaml` in their own gitops process if they prefer another way.

Reviewer 3: I'm ok with keeping it simple and use existing imageOverrides + Documentation.

   - TBD if operator yaml will be embedded in clusterctl or if it should be a

>      special artifact within the core provider repository.

Reviewer 3: I would say the operator should have its own image, so I'm +1 to remove this TBD

Author: @fabriziopandini So we are leaning towards the operator having its own image and its own yaml that will be specified as part of the cluster-api release artifacts. Correct?

I wanted to confirm because I wanted to add a note about "how" we check that the operator needs an upgrade. 
That is, does the user get a newer version of clusterctl which has a newer version of the operator yaml embedded within it.
OR
For example, when the user does `clusterctl upgrade plan`, will clusterctl by default check the Cluster API releases and see if there is a newer version of CAPI and hence a newer version of the operator yaml.
In this scenario for the air-gapped environment, clusterctl can see the providers list in `~/.cluster-api/clusterctl.yaml` and use whatever repo is specified there.

_or we check both_ 🤷  WDYT?

Reviewer 3: I'm assuming that the operator should be a separated package because we want to make it possible to use CAPI without the operator and/or without clusterctl, but might be worth to get other opinions here @vincepri @ncdc

 labels:
   provider-components: azure
 name: v0.4.8
 namespace: capz-system
data:
 components: |
 # components for v0.4.8 yaml goes here
 metadata: |
 # metadata information goes here
---
apiVersion: management.cluster.x-k8s.io/v1alpha1
kind: InfrastructureProvider
metadata:
 name: azure
 namespace: capz-system
spec:
 version: v0.4.9
 secretName: azure-variables
 fetchConfig:
   selector:
     matchLabels:
       provider-components: azure
```

### Risks and Mitigation

#### Error Handling & Logging

Currently, clusterctl provides quick feedback regarding required variables
etc. With the operator in place we’ll need to ensure that the error messages
and logs are easily available to the user to verify progress.

#### Extensibility Options

Currently, clusterctl has a few extensibility options.  For example,
clusterctl is built on-top of a library that can be leveraged to build other
tools.

It also exposes an interface for template processing if we choose to go a
different route from `envsubst`. This may prove to be challenging in the
context of the operator as this would mean a change to the operator
binary/image. We could introduce a new behavior or communication protocol or

> hooks for the operator to interact with the custom template processor. This

Reviewer 3: I would clearly state this is out of scope of the first iteration (might be worth to add a non-goal)

Author: Sounds good. I'll add to the non-goals that we are not supporting other template processors and will be using the default `envsubst` implementation.

could be configured similarly to the fetch config, with multiple options built
in.

#### Upgrade from v1alpha3 management cluster to v1apha4/operator cluster

Cluster API will provide instructions on how to upgrade from a v1alpha3
management cluster, created by clusterctl to the new v1alpha4 management
cluster. These operations could require manual actions.

Some of the actions are described below:
- Run webhooks as part of the main manager. See [issue

>   3822](https://github.com/kubernetes-sigs/cluster-api/issues/3822)

Reviewer 3: I would say that this paragraph should go into implementation detail.
Might be worth adding a note that further detail might be added later in the cycle when v1alpha4 details will be defined.

Instead, inside risk and mitigation, we should add a point documenting that as of today it is not possible to determine the exact upgrade sequence given that v1alpha4 is still under definition.

## Additional Details

> ### Test Plan

Reviewer 3: II think we should explicitly talk about e2e tests.
- existing E2E test will verify the new version of init and upgrade as soon as ready
- a new E2E test is required to verify v1alpha3 --> v1alpha4 upgrades

### Non-Goals/Future Work

- Deprecate the clusterctl CLI.
- Manage cluster templates using the operator.
- Implement an operator driven version of `clusterctl move`.
- Manage cert-manager using the operator.
- Support multiple installations of the same provider within a management
  cluster in light of [issue 3042] and [issue 3354].

> - Support multiple template processors.

Reviewer 3: I would add a non-goal to support the installation of v1alpha3 providers using the operator

apiVersion: management.cluster.x-k8s.io/v1alpha1
kind: InfrastructureProvider
metadata:

>  name: myaws

Reviewer 3: I will live aws here, I don't think we should encourage to use custom names even if in this case the fetch config is specified

Author: 👍  I'll fix this one to aws.

apiVersion: management.cluster.x-k8s.io/v1alpha1
kind: InfrastructureProvider
metadata:

>  name: myazure

Reviewer 3: I will live azure here, I don't think we should encourage to use custom names even if in this case the fetch config is specified

Author: Hmm...I wanted to show an example where if a different name other than the default expected one is provided, then the user needs to specify a fetch configuration.
